### PR TITLE
Update 'SQL Layer' with SIGMOD paper info

### PR DESCRIPTION
**Pull request by rmloveland**
Addresses #10794

Summary of changes:

- Update page overview with more detail

- Update parsing, physical planning, and logical planning sections

- Add a section on schema changes

- Various small copy edits throughout

Much of the language in these changes is taken from the SIGMOD paper
[1].  All credit goes to the paper's authors, any errors were introduced by me.

[1] https://www.cockroachlabs.com/blog/cockroachdb-sigmod-2020/

### CODE REVIEW

**Comment by rmloveland** [comment]
Hi Becca, Jordan, and Paul!

Asking for your reviews because this PR touches the respective sections of the SIGMOD paper of which you are the authors (AFAICT according to that repo's history?) , namely:

- logical planning (@rytaft )
- physical planning (@jordanlewis )
- schema changes (@pbardea )

would each of you mind looking at the updates to the respective sections in this PR?

add'l question: can these changes be back ported to the v20.1 and/or v20.2 docs?

also if you have any other general feedback I am happy to update stuff!

thanks!

**Review by pbardea — APPROVED** [review]
LGTM, the schema change section is general enough that it applies to all previous versions.

**Review by rytaft — APPROVED** [review]
Sorry for the delay! I was OOO last week.

:lgtm:, just a few nits below.

@jordanlewis is OOO this week, but I'm pretty sure the physical planning section is also good to go. Presumably you can make tweaks later if need be.

All of these changes you made also apply to 20.1 and 20.2, so feel free to backport.

> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/10984#-:-MgfJKKhBktsloFoGfv6:b-q0xk5l)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis and @rmloveland)

___
*[v21.1/architecture/sql-layer.md, line 7 at r1](https://reviewable.io/reviews/cockroachdb/docs/10984#-MgfEwU9ANluMYgKnjFU:-MgfEwU9ANluMYgKnjFV:b-r166c6) ([raw file](https://github.com/cockroachdb/docs/blob/4990712ff5e55080ce69683bb89a7e3e87339cd1/v21.1/architecture/sql-layer.md#L7)):*
> ```Markdown
> ---
> 
> The SQL layer of CockroachDB's architecture exposes SQL API to developers and converts high-level [SQL statements](../sql-statements.html) into low-level read and write requests to the underlying key-value store, which are passed to the [Transaction Layer](transaction-layer.html)
> ```

nit: exposes SQL API -> maybe "exposes a SQL API" ?

nit: missing period at end of paragraph
___
*[v21.1/architecture/sql-layer.md, line 14 at r1](https://reviewable.io/reviews/cockroachdb/docs/10984#-MgfFZU262jerFS2Z9Ny:-MgfFZU36RD-tfnyMhOf:bg9soln) ([raw file](https://github.com/cockroachdb/docs/blob/4990712ff5e55080ce69683bb89a7e3e87339cd1/v21.1/architecture/sql-layer.md#L14)):*
> ```Markdown
> - [Parser](#parsing), which converts SQL text into an abstract syntax tree (AST).
> - [Cost-based optimizer](#logical-planning), which converts the AST into an optimized logical query plan.
> - [SQL execution engine](#physical-planning), which converts the logical query plan into a physical query plan for execution by one or more nodes in the cluster.
> ```

I think you need to split this last bullet point into two. Something like this:
- [Physical planner](#physical-planning), which converts the logical query plan into a physical query plan for execution by one or more nodes in the cluster
- [SQL execution engine](#query-execution), which executes the physical plan by making read and write requests to the underlying key-value store
___
*[v21.1/architecture/sql-layer.md, line 41 at r1](https://reviewable.io/reviews/cockroachdb/docs/10984#-MgfH51y7LJKPpfu2yEC:-MgfH51y7LJKPpfu2yED:b2pw114) ([raw file](https://github.com/cockroachdb/docs/blob/4990712ff5e55080ce69683bb89a7e3e87339cd1/v21.1/architecture/sql-layer.md#L41)):*
> ```Markdown
> Developers experience data stored in CockroachDB as a relational structure comprised of rows and columns. Sets of rows and columns are further organized into [tables](../show-tables.html). Collections of tables are then organized into [databases](../show-databases.html). A CockroachDB cluster can contain many databases.
> 
> Because of this structure, CockroachDB provides typical relational features like [constraints](../constraints.html) (e.g., [foreign keys](../foreign-key.html)). These features mean that application developers can trust that the database will ensure consistent structuring of the application's data; data validation doesn't need to be built into the application logic separately.
> ```

nit: it's a bit weird to say we provide these features "Because of this structure". Maybe just get rid of that phrase?

**Review by rmloveland — COMMENTED** [review]
Thanks Becca and Paul for the reviews!

Becca, thanks for the feedback items - made the various requested changes.  Removing Jordan per your comment, agree can tweak later as needed

Will pass on for docs team reviewer and once it's ready, will back port to v20.1 and v20.2

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/10984#-:-MggTVL_AXTlvVbPR0VV:b-ojku3s)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis and @rytaft)

___
*[v21.1/architecture/sql-layer.md, line 7 at r1](https://reviewable.io/reviews/cockroachdb/docs/10984#-MgfEwU9ANluMYgKnjFU:-MggTlbm2vXN6sJj0Ppx:bj3eodx) ([raw file](https://github.com/cockroachdb/docs/blob/4990712ff5e55080ce69683bb89a7e3e87339cd1/v21.1/architecture/sql-layer.md#L7)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: exposes SQL API -> maybe "exposes a SQL API" ?

nit: missing period at end of paragraph
</blockquote></details>

Fixed, thanks!
___
*[v21.1/architecture/sql-layer.md, line 14 at r1](https://reviewable.io/reviews/cockroachdb/docs/10984#-MgfFZU262jerFS2Z9Ny:-MggVE_e9FF_UMlpOO0L:b-mrlffj) ([raw file](https://github.com/cockroachdb/docs/blob/4990712ff5e55080ce69683bb89a7e3e87339cd1/v21.1/architecture/sql-layer.md#L14)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I think you need to split this last bullet point into two. Something like this:
- [Physical planner](#physical-planning), which converts the logical query plan into a physical query plan for execution by one or more nodes in the cluster
- [SQL execution engine](#query-execution), which executes the physical plan by making read and write requests to the underlying key-value store
</blockquote></details>

Thanks!  totes went with that
___
*[v21.1/architecture/sql-layer.md, line 41 at r1](https://reviewable.io/reviews/cockroachdb/docs/10984#-MgfH51y7LJKPpfu2yEC:-MggVPv48xtTMUy0uL2P:b-k7mx83) ([raw file](https://github.com/cockroachdb/docs/blob/4990712ff5e55080ce69683bb89a7e3e87339cd1/v21.1/architecture/sql-layer.md#L41)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: it's a bit weird to say we provide these features "Because of this structure". Maybe just get rid of that phrase?
</blockquote></details>

Deleted!

**Review by rytaft — APPROVED** [review]
> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/10984)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @taroface)

**Review by taroface — APPROVED** [review]
LGTM! Just nits and style questions.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/10984#-:-MggtXIV8uDRbz14tlPY:b-jraecd)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rmloveland)

___
*[v21.1/architecture/sql-layer.md, line 7 at r2](https://reviewable.io/reviews/cockroachdb/docs/10984#-MggqFuJ3XRyGeOP53YU:-MggqFuJ3XRyGeOP53YV:bkd9zp6) ([raw file](https://github.com/cockroachdb/docs/blob/886a6e7c9fd0855e6f781a9eb83e21336be5dd4a/v21.1/architecture/sql-layer.md#L7)):*
> ```Markdown
> ---
> 
> The SQL layer of CockroachDB's architecture exposes a SQL API to developers and converts high-level [SQL statements](../sql-statements.html) into low-level read and write requests to the underlying key-value store, which are passed to the [Transaction Layer](transaction-layer.html).
> ```

Formatting nit: Transaction Layer -> transaction layer
___
*[v21.1/architecture/sql-layer.md, line 56 at r2](https://reviewable.io/reviews/cockroachdb/docs/10984#-MggqvaZ8aHLfqX2a2q4:-MggqvaZ8aHLfqX2a2q5:b-yefghe) ([raw file](https://github.com/cockroachdb/docs/blob/886a6e7c9fd0855e6f781a9eb83e21336be5dd4a/v21.1/architecture/sql-layer.md#L56)):*
> ```
> it [parses the statement](#parsing), [creates an optimized logical query plan](#logical-planning) which is further translated into a [physical query plan](#physical-planning).
> ```

Feels like this should be "parses the statement and creates an optimized logical query plan that is further translated..."
___
*[v21.1/architecture/sql-layer.md, line 64 at r2](https://reviewable.io/reviews/cockroachdb/docs/10984#-MggrqZqD8iv5zmCT66d:-MggrqZqD8iv5zmCT66e:b1m3iez) ([raw file](https://github.com/cockroachdb/docs/blob/886a6e7c9fd0855e6f781a9eb83e21336be5dd4a/v21.1/architecture/sql-layer.md#L64)):*
> ```
> _logical planning_
> ```

Do you want to use asterisks here for consistency with the italicized words below?
___
*[v21.1/architecture/sql-layer.md, line 69 at r2](https://reviewable.io/reviews/cockroachdb/docs/10984#-MggrQPsE4nB97VrIgY8:-MggrQPsE4nB97VrIgY9:b-q4e6e0) ([raw file](https://github.com/cockroachdb/docs/blob/886a6e7c9fd0855e6f781a9eb83e21336be5dd4a/v21.1/architecture/sql-layer.md#L69)):*
> ```
> e.g.
> ```

Add a comma after "e.g."

Also, suggest rewording: "e.g., by replacing`0.6 + 0.4` with `1.0`"
___
*[v21.1/architecture/sql-layer.md, line 70 at r2](https://reviewable.io/reviews/cockroachdb/docs/10984#-MggrXy20nrwH_3ZT_Yo:-MggrXy20nrwH_3ZT_Yp:b24zgai) ([raw file](https://github.com/cockroachdb/docs/blob/886a6e7c9fd0855e6f781a9eb83e21336be5dd4a/v21.1/architecture/sql-layer.md#L70)):*
> ```
> for example
> ```

Make this "e.g.," for consistency?
___
*[v21.1/architecture/sql-layer.md, line 85 at r2](https://reviewable.io/reviews/cockroachdb/docs/10984#-MggsARGDSTnK0579FMf:-MggsARHDtmbm9_ScJfp:b-i6i440) ([raw file](https://github.com/cockroachdb/docs/blob/886a6e7c9fd0855e6f781a9eb83e21336be5dd4a/v21.1/architecture/sql-layer.md#L85)):*
> ```
> _SQL operators_
> ```

Ditto re: asterisks
___
*[v21.1/architecture/sql-layer.md, line 87 at r2](https://reviewable.io/reviews/cockroachdb/docs/10984#-MggsIVr9jhku0a78hhe:-MggsIVr9jhku0a78hhf:b-sub1s3) ([raw file](https://github.com/cockroachdb/docs/blob/886a6e7c9fd0855e6f781a9eb83e21336be5dd4a/v21.1/architecture/sql-layer.md#L87)):*
> ```
> key space
> ```

This might be a question for #docs-style, but I also see this as `key-space` and `keyspace` in other docs (including architecture docs). Which usage is best?
___
*[v21.1/architecture/sql-layer.md, line 91 at r2](https://reviewable.io/reviews/cockroachdb/docs/10984#-MggsZ480ptuzbJKzrZS:-MggsZ480ptuzbJKzrZT:b-e7z2da) ([raw file](https://github.com/cockroachdb/docs/blob/886a6e7c9fd0855e6f781a9eb83e21336be5dd4a/v21.1/architecture/sql-layer.md#L91)):*
> ```
> _TableReader_
> ```

Should this be in code backticks?
___
*[v21.1/architecture/sql-layer.md, line 91 at r2](https://reviewable.io/reviews/cockroachdb/docs/10984#-Mggscue9V8fN9E489uR:-Mggscue9V8fN9E489uS:bku2km9) ([raw file](https://github.com/cockroachdb/docs/blob/886a6e7c9fd0855e6f781a9eb83e21336be5dd4a/v21.1/architecture/sql-layer.md#L91)):*
> ```
> TableReaders
> ```

Ditto
___
*[v21.1/architecture/sql-layer.md, line 91 at r2](https://reviewable.io/reviews/cockroachdb/docs/10984#-MggshfeEq6qlsMOMyl8:-MggshfeEq6qlsMOMyl9:bko636u) ([raw file](https://github.com/cockroachdb/docs/blob/886a6e7c9fd0855e6f781a9eb83e21336be5dd4a/v21.1/architecture/sql-layer.md#L91)):*
> ```
> close
> ```

closely? Actually not sure, I'm a bit tired -_-
___
*[v21.1/architecture/sql-layer.md, line 144 at r2](https://reviewable.io/reviews/cockroachdb/docs/10984#-Mggt4kEEksve_NnEWw6:-Mggt4kEEksve_NnEWw7:b-2x6osl) ([raw file](https://github.com/cockroachdb/docs/blob/886a6e7c9fd0855e6f781a9eb83e21336be5dd4a/v21.1/architecture/sql-layer.md#L144)):*
> ```
> _Online, Asynchronous Schema Change in F1_
> ```

Another style question, but I think paper titles might want to be in quotation marks rather than ital.

**Review by rmloveland — COMMENTED** [review]
Thanks for the review Ryan!!!1! I think almost everything is covered, modulo a couple style things we can fix in a followup after deciding in #docs-style !

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/10984#-:-MglIZ9pCypq1PEZLISf:biq7bpi)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @taroface)

___
*[v21.1/architecture/sql-layer.md, line 7 at r2](https://reviewable.io/reviews/cockroachdb/docs/10984#-MggqFuJ3XRyGeOP53YU:-MglIcon3OFOOzNOmoGE:bt2vr0y) ([raw file](https://github.com/cockroachdb/docs/blob/886a6e7c9fd0855e6f781a9eb83e21336be5dd4a/v21.1/architecture/sql-layer.md#L7)):*
<details><summary><i>Previously, taroface (Ryan Kuo) wrote…</i></summary><blockquote>

Formatting nit: Transaction Layer -> transaction layer
</blockquote></details>

Fixed!
___
*[v21.1/architecture/sql-layer.md, line 56 at r2](https://reviewable.io/reviews/cockroachdb/docs/10984#-MggqvaZ8aHLfqX2a2q4:-MglIl4r7_8Hx8vikTVV:blft88q) ([raw file](https://github.com/cockroachdb/docs/blob/886a6e7c9fd0855e6f781a9eb83e21336be5dd4a/v21.1/architecture/sql-layer.md#L56)):*
<details><summary><i>Previously, taroface (Ryan Kuo) wrote…</i></summary><blockquote>

> ```
> it [parses the statement](#parsing), [creates an optimized logical query plan](#logical-planning) which is further translated into a [physical query plan](#physical-planning).
> ```

Feels like this should be "parses the statement and creates an optimized logical query plan that is further translated..."
</blockquote></details>

Updated!  PS kudos on knowing which vs that, I literally have no idea and constantly switch them.  Guess I should finally learn this .... 
___
*[v21.1/architecture/sql-layer.md, line 64 at r2](https://reviewable.io/reviews/cockroachdb/docs/10984#-MggrqZqD8iv5zmCT66d:-MglJ5UGA2dfp5dXthmc:b-hstd8a) ([raw file](https://github.com/cockroachdb/docs/blob/886a6e7c9fd0855e6f781a9eb83e21336be5dd4a/v21.1/architecture/sql-layer.md#L64)):*
<details><summary><i>Previously, taroface (Ryan Kuo) wrote…</i></summary><blockquote>

> ```
> _logical planning_
> ```

Do you want to use asterisks here for consistency with the italicized words below?
</blockquote></details>

Sure - done!
___
*[v21.1/architecture/sql-layer.md, line 69 at r2](https://reviewable.io/reviews/cockroachdb/docs/10984#-MggrQPsE4nB97VrIgY8:-MglJFjQ1G3lUIEjRQLq:bx4nh5p) ([raw file](https://github.com/cockroachdb/docs/blob/886a6e7c9fd0855e6f781a9eb83e21336be5dd4a/v21.1/architecture/sql-layer.md#L69)):*
<details><summary><i>Previously, taroface (Ryan Kuo) wrote…</i></summary><blockquote>

> ```
> e.g.
> ```

Add a comma after "e.g."

Also, suggest rewording: "e.g., by replacing`0.6 + 0.4` with `1.0`"
</blockquote></details>

Done and done!
___
*[v21.1/architecture/sql-layer.md, line 70 at r2](https://reviewable.io/reviews/cockroachdb/docs/10984#-MggrXy20nrwH_3ZT_Yo:-MglJJC65TZmBQ7yFnuZ:b-896fja) ([raw file](https://github.com/cockroachdb/docs/blob/886a6e7c9fd0855e6f781a9eb83e21336be5dd4a/v21.1/architecture/sql-layer.md#L70)):*
<details><summary><i>Previously, taroface (Ryan Kuo) wrote…</i></summary><blockquote>

> ```
> for example
> ```

Make this "e.g.," for consistency?
</blockquote></details>

Done!
___
*[v21.1/architecture/sql-layer.md, line 85 at r2](https://reviewable.io/reviews/cockroachdb/docs/10984#-MggsARGDSTnK0579FMf:-MglJOGQASN9cSoxwupE:b-896fja) ([raw file](https://github.com/cockroachdb/docs/blob/886a6e7c9fd0855e6f781a9eb83e21336be5dd4a/v21.1/architecture/sql-layer.md#L85)):*
<details><summary><i>Previously, taroface (Ryan Kuo) wrote…</i></summary><blockquote>

> ```
> _SQL operators_
> ```

Ditto re: asterisks
</blockquote></details>

Done!
___
*[v21.1/architecture/sql-layer.md, line 87 at r2](https://reviewable.io/reviews/cockroachdb/docs/10984#-MggsIVr9jhku0a78hhe:-MglJWkrECyeUa1XXn4y:bz23ef5) ([raw file](https://github.com/cockroachdb/docs/blob/886a6e7c9fd0855e6f781a9eb83e21336be5dd4a/v21.1/architecture/sql-layer.md#L87)):*
<details><summary><i>Previously, taroface (Ryan Kuo) wrote…</i></summary><blockquote>

> ```
> key space
> ```

This might be a question for #docs-style, but I also see this as `key-space` and `keyspace` in other docs (including architecture docs). Which usage is best?
</blockquote></details>

oh noes we def have all 3!!!!!

I may have to punt this to another PR.  Filed  #11029 
___
*[v21.1/architecture/sql-layer.md, line 91 at r2](https://reviewable.io/reviews/cockroachdb/docs/10984#-MggsZ480ptuzbJKzrZS:-MglKYg6DrJLjV1h5mim:b-h9ntbd) ([raw file](https://github.com/cockroachdb/docs/blob/886a6e7c9fd0855e6f781a9eb83e21336be5dd4a/v21.1/architecture/sql-layer.md#L91)):*
<details><summary><i>Previously, taroface (Ryan Kuo) wrote…</i></summary><blockquote>

> ```
> _TableReader_
> ```

Should this be in code backticks?
</blockquote></details>

I went with italics since that is what was used in the SIGMOD paper.  I think italics may be better also since we are somewhat trying to get away from mentioning explicit names of things from the code in the architecture docs (since the specifics can change etc.)
___
*[v21.1/architecture/sql-layer.md, line 91 at r2](https://reviewable.io/reviews/cockroachdb/docs/10984#-Mggscue9V8fN9E489uR:-MglKmyF6j8FhQzDEf32:b-34vmyd) ([raw file](https://github.com/cockroachdb/docs/blob/886a6e7c9fd0855e6f781a9eb83e21336be5dd4a/v21.1/architecture/sql-layer.md#L91)):*
<details><summary><i>Previously, taroface (Ryan Kuo) wrote…</i></summary><blockquote>

> ```
> TableReaders
> ```

Ditto
</blockquote></details>

See above
___
*[v21.1/architecture/sql-layer.md, line 91 at r2](https://reviewable.io/reviews/cockroachdb/docs/10984#-MggshfeEq6qlsMOMyl8:-MglKoBiEhi0HAP5e9gF:b-dqchk8) ([raw file](https://github.com/cockroachdb/docs/blob/886a6e7c9fd0855e6f781a9eb83e21336be5dd4a/v21.1/architecture/sql-layer.md#L91)):*
<details><summary><i>Previously, taroface (Ryan Kuo) wrote…</i></summary><blockquote>

> ```
> close
> ```

closely? Actually not sure, I'm a bit tired -_-
</blockquote></details>

I think "as close .. as possible" is also OK? (I think? I am also tired)
___
*[v21.1/architecture/sql-layer.md, line 144 at r2](https://reviewable.io/reviews/cockroachdb/docs/10984#-Mggt4kEEksve_NnEWw6:-MglM-1d598fTUxf6Aat:brh01bd) ([raw file](https://github.com/cockroachdb/docs/blob/886a6e7c9fd0855e6f781a9eb83e21336be5dd4a/v21.1/architecture/sql-layer.md#L144)):*
<details><summary><i>Previously, taroface (Ryan Kuo) wrote…</i></summary><blockquote>

> ```
> _Online, Asynchronous Schema Change in F1_
> ```

Another style question, but I think paper titles might want to be in quotation marks rather than ital.
</blockquote></details>

IDK - let's discuss in #docs-style !
